// rc-segmented/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var React = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _rcMotion = _interopRequireDefault(require("rc-motion"));

var _useMergedState3 = _interopRequireDefault(require("rc-util/lib/hooks/useMergedState"));

var _ref3 = require("rc-util/lib/ref");

var _omit = _interopRequireDefault(require("rc-util/lib/omit"));

var _excluded = ["prefixCls", "direction", "options", "disabled", "defaultValue", "value", "onChange", "className", "motionName"];

function getValidTitle(option) {
  if (typeof option.title !== 'undefined') {
    return option.title;
  } // read `label` when title is `undefined`


  if ((0, _typeof2.default)(option.label) !== 'object') {
    var _option$label;

    return (_option$label = option.label) === null || _option$label === void 0 ? void 0 : _option$label.toString();
  }
}

function normalizeOptions(options) {
  return options.map(function (option) {
    if ((0, _typeof2.default)(option) === 'object' && option !== null) {
      var validTitle = getValidTitle(option);
      return (0, _objectSpread2.default)((0, _objectSpread2.default)({}, option), {}, {
        title: validTitle
      });
    }

    return {
      label: option === null || option === void 0 ? void 0 : option.toString(),
      title: option === null || option === void 0 ? void 0 : option.toString(),
      value: option
    };
  });
}

var calcThumbStyle = function calcThumbStyle(targetElement) {
  return {
    transform: "translateX(".concat(targetElement.offsetLeft, "px)"),
    width: targetElement.clientWidth
  };
};

var InternalSegmentedOption = function InternalSegmentedOption(_ref) {
  var prefixCls = _ref.prefixCls,
      className = _ref.className,
      disabled = _ref.disabled,
      checked = _ref.checked,
      label = _ref.label,
      title = _ref.title,
      value = _ref.value,
      onChange = _ref.onChange;

  var handleChange = function handleChange(event) {
    if (disabled) {
      return;
    }

    onChange(event, value);
  };

  return /*#__PURE__*/React.createElement("label", {
    className: (0, _classnames.default)(className, (0, _defineProperty2.default)({}, "".concat(prefixCls, "-item-disabled"), disabled))
  }, /*#__PURE__*/React.createElement("input", {
    className: "".concat(prefixCls, "-item-input"),
    type: "radio",
    disabled: disabled,
    checked: checked,
    onChange: handleChange
  }), /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-item-label"),
    title: title
  }, label));
};

var Segmented = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _segmentedOptions$, _classNames2;

  var _props$prefixCls = props.prefixCls,
      prefixCls = _props$prefixCls === void 0 ? 'rc-segmented' : _props$prefixCls,
      direction = props.direction,
      options = props.options,
      disabled = props.disabled,
      defaultValue = props.defaultValue,
      value = props.value,
      onChange = props.onChange,
      _props$className = props.className,
      className = _props$className === void 0 ? '' : _props$className,
      _props$motionName = props.motionName,
      motionName = _props$motionName === void 0 ? 'thumb-motion' : _props$motionName,
      restProps = (0, _objectWithoutProperties2.default)(props, _excluded);
  var containerRef = React.useRef(null);
  var mergedRef = (0, _ref3.composeRef)(containerRef, ref);
  var thumbMoveStatus = React.useRef({
    from: null,
    to: null
  });
  var segmentedOptions = React.useMemo(function () {
    return normalizeOptions(options);
  }, [options]);

  var _useMergedState = (0, _useMergedState3.default)((_segmentedOptions$ = segmentedOptions[0]) === null || _segmentedOptions$ === void 0 ? void 0 : _segmentedOptions$.value, {
    value: props.value,
    defaultValue: defaultValue
  }),
      _useMergedState2 = (0, _slicedToArray2.default)(_useMergedState, 2),
      selected = _useMergedState2[0],
      setSelected = _useMergedState2[1];

  var _React$useState = React.useState(selected),
      _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
      visualSelected = _React$useState2[0],
      setVisualSelected = _React$useState2[1];

  var _React$useState3 = React.useState(false),
      _React$useState4 = (0, _slicedToArray2.default)(_React$useState3, 2),
      thumbShow = _React$useState4[0],
      setThumbShow = _React$useState4[1];

  var doThumbAnimation = React.useCallback(function (selectedValue) {
    var _containerRef$current;

    var segmentedItemIndex = segmentedOptions.findIndex(function (n) {
      return n.value === selectedValue;
    });

    if (segmentedItemIndex < 0) {
      return;
    } // find target element


    var toElement = (_containerRef$current = containerRef.current) === null || _containerRef$current === void 0 ? void 0 : _containerRef$current.querySelector(".".concat(prefixCls, "-item:nth-child(").concat(segmentedItemIndex + 1, ")"));

    if (toElement) {
      var _containerRef$current2;

      // find source element
      var fromElement = (_containerRef$current2 = containerRef.current) === null || _containerRef$current2 === void 0 ? void 0 : _containerRef$current2.querySelector(".".concat(prefixCls, "-item-selected"));

      if (fromElement && toElement && thumbMoveStatus.current) {
        // calculate for thumb moving animation
        thumbMoveStatus.current.from = calcThumbStyle(fromElement);
        thumbMoveStatus.current.to = calcThumbStyle(toElement); // trigger css-motion starts

        setThumbShow(true);
      }
    }
  }, [prefixCls, segmentedOptions]); // get latest version of `visualSelected`

  var latestVisualSelected = React.useRef(visualSelected);
  React.useEffect(function () {
    latestVisualSelected.current = visualSelected;
  });
  React.useEffect(function () {
    // Syncing `visualSelected` when `selected` changed
    // and do thumb animation
    if ((typeof selected === 'string' || typeof selected === 'number') && selected !== latestVisualSelected.current) {
      doThumbAnimation(selected);
    }
  }, [selected]);

  var handleChange = function handleChange(event, val) {
    if (disabled) {
      return;
    }

    setSelected(val);
    onChange === null || onChange === void 0 ? void 0 : onChange(val);
  }; // --- motion event handlers for thumb move


  var handleThumbEnterStart = function handleThumbEnterStart() {
    var fromStyle = thumbMoveStatus.current.from;

    if (fromStyle) {
      setVisualSelected(undefined);
      return fromStyle;
    }
  };

  var handleThumbEnterActive = function handleThumbEnterActive() {
    var toStyle = thumbMoveStatus.current.to;

    if (toStyle) {
      return toStyle;
    }
  };

  var handleThumbEnterEnd = function handleThumbEnterEnd() {
    setThumbShow(false);
    setVisualSelected(selected);

    if (thumbMoveStatus.current) {
      thumbMoveStatus.current = {
        from: null,
        to: null
      };
    }
  };

  var divProps = (0, _omit.default)(restProps, ['children']);
  return /*#__PURE__*/React.createElement("div", (0, _objectSpread2.default)((0, _objectSpread2.default)({}, divProps), {}, {
    className: (0, _classnames.default)(prefixCls, (_classNames2 = {}, (0, _defineProperty2.default)(_classNames2, "".concat(prefixCls, "-rtl"), direction === 'rtl'), (0, _defineProperty2.default)(_classNames2, "".concat(prefixCls, "-disabled"), disabled), _classNames2), className),
    ref: mergedRef
  }), /*#__PURE__*/React.createElement(_rcMotion.default, {
    visible: thumbShow,
    motionName: "".concat(prefixCls, "-").concat(motionName),
    motionDeadline: 300,
    onEnterStart: handleThumbEnterStart,
    onEnterActive: handleThumbEnterActive,
    onEnterEnd: handleThumbEnterEnd
  }, function (_ref2) {
    var motionClassName = _ref2.className,
        motionStyle = _ref2.style;
    return /*#__PURE__*/React.createElement("div", {
      style: motionStyle,
      className: (0, _classnames.default)("".concat(prefixCls, "-thumb"), motionClassName)
    });
  }), segmentedOptions.map(function (segmentedOption) {
    return /*#__PURE__*/React.createElement(InternalSegmentedOption, (0, _objectSpread2.default)({
      key: segmentedOption.value,
      prefixCls: prefixCls,
      className: (0, _classnames.default)(segmentedOption.className, "".concat(prefixCls, "-item"), (0, _defineProperty2.default)({}, "".concat(prefixCls, "-item-selected"), segmentedOption.value === visualSelected)),
      checked: segmentedOption.value === selected,
      onChange: handleChange
    }, segmentedOption));
  }));
});
Segmented.displayName = 'Segmented';
Segmented.defaultProps = {
  options: []
};
var _default = Segmented;
exports.default = _default;
